src/transaction/log_page_buffer.c

```diff
@@ -7653,10 +7653,12 @@ logpb_checkpoint_trantable (THREAD_ENTRY * const thread_p)
 
   log_write_metalog_to_file (false);
 
-  // function explicitly needs to be called in critical section-free context
-  LOG_CS_EXIT (thread_p);
-  logpb_flush_pages (thread_p, &trantable_checkpoint_lsa);
-  LOG_CS_ENTER (thread_p);
+  /* NOTE. in development (TODO)
+   * If the log records up to trantable_checkpoint_lsa is not stored on PS,
+   * it will fail to do recovery with PS's which have lagged behind the snapshot LSA.
+   * In the near future, ATS will restore with the checkpoint info from the log in PS and it will work well.
+   */
+  logpb_flush_pages_direct (thread_p);
```

Comment:
I looked and debugged the code with and without the change.
I think this change is not enough.

Upon {{cubrid server stop...}} the ATS will still hang in the HA delay daemon with the stack:
```
Thread 21 (Thread 0x7fa50cff9700 (LWP 51253) "cub_server"):
#0  0x00007fa538fdba35 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00007fa538b75aec in std::condition_variable::wait(std::unique_lock<std::mutex>&) () from /lib64/libstdc++.so.6
#2  0x00007fa539eccacf in log_global::wait_for_ps_flushed_lsa (this=0x7fa53a78eea0 <log_Gl>, flush_lsa=...) at /home/qa/cubrid-src/src/transaction/log_global.c:137
#3  0x00007fa539f0184f in logpb_flush_pages (thread_p=0x7fa53a952c98, flush_lsa=0x7fa50cff8c08) at /home/qa/cubrid-src/src/transaction/log_page_buffer.c:4512
#4  0x00007fa539ed46d2 in log_append_ha_server_state (thread_p=0x7fa53a952c98, state=1) at /home/qa/cubrid-src/src/transaction/log_manager.c:3404
#5  0x00007fa539ee2dbb in log_check_ha_delay_info_execute (thread_ref=...) at /home/qa/cubrid-src/src/transaction/log_manager.c:10695
#6  0x00007fa539d74b15 in std::_Function_handler<void (cubthread::entry&), void (*)(cubthread::entry&)>::_M_invoke(std::_Any_data const&, cubthread::entry&) (__functor=..., __args#0=...) at /opt/rh/devtoolset-9/root/usr/include/c++/9/bits/std_function.h:300
#7  0x00007fa539cbd285 in std::function<void (cubthread::entry&)>::operator()(cubthread::entry&) const (this=0x29dec28, __args#0=...) at /opt/rh/devtoolset-9/root/usr/include/c++/9/bits/std_function.h:688
#8  0x00007fa539cbccad in cubthread::callable_task<cubthread::entry>::execute (this=0x29dec20, ctx=...) at /home/qa/cubrid-src/src/thread/thread_task.hpp:104
#9  0x00007fa539e68244 in cubthread::daemon::loop_with_context<cubthread::entry> (daemon_arg=0x29dec70, context_manager_arg=0x142b700, exec_arg=0x29dec20, name=0x29dee78 "log_check_ha_delay_info") at /home/qa/cubrid-src/src/thread/thread_daemon.hpp:183
#10 0x00007fa539e6b1e9 in std::__invoke_impl<void, void (*)(cubthread::daemon*, cubthread::context_manager<cubthread::entry>*, cubthread::task<cubthread::entry>*, char const*), cubthread::daemon*, cubthread::context_manager<cubthread::entry>*, cubthread::task<cubthread::entry>*, char const*> (__f=@0x29deee8: 0x7fa539e6816e <cubthread::daemon::loop_with_context<cubthread::entry>(cubthread::daemon*, cubthread::context_manager<cubthread::entry>*, cubthread::task<cubthread::entry>*, char const*)>) at /opt/rh/devtoolset-9/root/usr/include/c++/9/bits/invoke.h:60
#11 0x00007fa539e6b0b2 in std::__invoke<void (*)(cubthread::daemon*, cubthread::context_manager<cubthread::entry>*, cubthread::task<cubthread::entry>*, char const*), cubthread::daemon*, cubthread::context_manager<cubthread::entry>*, cubthread::task<cubthread::entry>*, char const*> (__fn=@0x29deee8: 0x7fa539e6816e <cubthread::daemon::loop_with_context<cubthread::entry>(cubthread::daemon*, cubthread::context_manager<cubthread::entry>*, cubthread::task<cubthread::entry>*, char const*)>) at /opt/rh/devtoolset-9/root/usr/include/c++/9/bits/invoke.h:95
#12 0x00007fa539e6af6d in std::thread::_Invoker<std::tuple<void (*)(cubthread::daemon*, cubthread::context_manager<cubthread::entry>*, cubthread::task<cubthread::entry>*, char const*), cubthread::daemon*, cubthread::context_manager<cubthread::entry>*, cubthread::task<cubthread::entry>*, char const*> >::_M_invoke<0ul, 1ul, 2ul, 3ul, 4ul> (this=0x29deec8) at /opt/rh/devtoolset-9/root/usr/include/c++/9/thread:244
#13 0x00007fa539e6aed4 in std::thread::_Invoker<std::tuple<void (*)(cubthread::daemon*, cubthread::context_manager<cubthread::entry>*, cubthread::task<cubthread::entry>*, char const*), cubthread::daemon*, cubthread::context_manager<cubthread::entry>*, cubthread::task<cubthread::entry>*, char const*> >::operator() (this=0x29deec8) at /opt/rh/devtoolset-9/root/usr/include/c++/9/thread:251
#14 0x00007fa539e6aeb8 in std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (*)(cubthread::daemon*, cubthread::context_manager<cubthread::entry>*, cubthread::task<cubthread::entry>*, char const*), cubthread::daemon*, cubthread::context_manager<cubthread::entry>*, cubthread::task<cubthread::entry>*, char const*> > >::_M_run (this=0x29deec0) at /opt/rh/devtoolset-9/root/usr/include/c++/9/thread:195
#15 0x00007fa53a25d1c0 in execute_native_thread_routine () from /home/qa/CUBRID/lib/libcubrid.so.12.0
#16 0x00007fa538fd7ea5 in start_thread () from /lib64/libpthread.so.0
#17 0x00007fa5382dcb0d in clone () from /lib64/libc.so.6

Thread 1 (Thread 0x7fa53a9f0780 (LWP 51231) "cub_server"):
#0  0x00007fa538fd9017 in pthread_join () from /lib64/libpthread.so.0
#1  0x00007fa538b790f7 in std::thread::join() () from /lib64/libstdc++.so.6
#2  0x00007fa539e5d12e in cubthread::daemon::stop_execution (this=0x29dec70) at /home/qa/cubrid-src/src/thread/thread_daemon.cpp:97
#3  0x00007fa539e66b4f in cubthread::manager::destroy_and_untrack_resource<cubthread::daemon> (this=0x142b630, tracker=..., res=@0x7fa53a7a17f8: 0x29dec70, entries_count=1) at /home/qa/cubrid-src/src/thread/thread_manager.cpp:247
#4  0x00007fa539e656bf in cubthread::manager::destroy_daemon (this=0x142b630, daemon_arg=@0x7fa53a7a17f8: 0x29dec70) at /home/qa/cubrid-src/src/thread/thread_manager.cpp:365
#5  0x00007fa539ed1ec8 in log_stop_ha_delay_registration () at /home/qa/cubrid-src/src/transaction/log_manager.c:2046
#6  0x00007fa539e80a99 in xboot_shutdown_server (thread_p=@0x7fffcd781b38: 0x142b720, is_er_final=ER_THREAD_FINAL) at /home/qa/cubrid-src/src/transaction/boot_sr.c:3167
#7  0x00007fa5399e2a40 in net_server_start (thread_p=0x142b720, server_name=0x7fffcd783aba "scaldb") at /home/qa/cubrid-src/src/communication/network_sr.c:1151
#8  0x000000000040c078 in main (argc=2, argv=0x7fffcd781c78) at /home/qa/cubrid-src/src/executables/server.c:434
```

And the `ha_delay` daemon we keep for various purposes:
- one is that to measure replication delay (but this is not so important and we could circumvent it some other way)
- another one is that I still think we have an unsolved issue regarding the redo_lsa in the page server log replicator (see http://jira.cubrid.org/browse/LETS-656 )

Reply:
Did we already talk about it? I feel like a dejavu.
Regarding the ha_delay log, I don't think we need to take it into considerations for this problem because
1. The ha-delay log record is for old-HA, and I think I was told it would be removed. As you said the delay will probably be measured in another way.
2. About the issue, LETS-656, I can't find the reason why log records must keep to be generated continuously. As the issue is saying, the hang occurs if it's not, but I think it's kind of bug we should address.

Besides, if (1) and (2) are needed, I doubt they need to meet the quorum as the ha_delay does.

Reply:
> As the issue is saying, the hang occurs if it's not, but I think it's kind of bug we should address.

Indeed.

Ok, Let's get on with this even though I have doubts:
- doubts that we need to remove the quorum confirmation for checkpoint
- I think some of the existing tests might fail - if they don't it means that we don't have sufficiently covering tests or we need more stress tests (as described in http://jira.cubrid.org/browse/LETS-656 )